plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,729 @@
+
+  // sweep to clear duplicates, last one wins
+  for (size_t index = 0; index < (loaded.size() - 1); ++index) {
+    if (loaded[index].regex_text == loaded[index + 1].regex_text) {
+      loaded[index] = Invalidate{};
+    }
+  }
+
+  if (!this->invalidate_vec || this->invalidate_vec->empty()) {
+    DEBUG_LOG("Installing fresh rules");
+    *rules = std::move(loaded);
+    return true;
+  }
+
+  DEBUG_LOG("Merging new config");
+
+  // merge loaded and current rule set
+  auto const &cur = *(this->invalidate_vec);
+  auto itload     = loaded.begin();
+  auto itcur      = cur.cbegin();
+
+  bool changed = false;
+
+  // reimplementation of std::set_union
+  while (cur.cend() != itcur) {
+    if (loaded.cend() == itload) {
+      std::copy(itcur, cur.cend(), rules->end());
+      break;
+    }
+
+    // fast forward over cleared items
+    while (!itload->is_valid()) {
+      DEBUG_LOG("Skipping cleared duplicate rule");
+      ++itload;
+      TSAssert(loaded.end() != itload); // last item will always be valid.
+    }
+
+    int const cmp = itload->regex_text.compare(itcur->regex_text);
+    if (cmp < 0) {
+      if (timenow < itload->expiry) {
+        DEBUG_LOG("Adding new rule: '%s'", itload->regex_text.c_str());
+        rules->push_back(std::move(*itload));
+        changed = true;
+      } else {
+        DEBUG_LOG("Not adding new expired rule: '%s'", itload->regex_text.c_str());
+      }
+      ++itload;
+    } else if (0 < cmp) {
+      DEBUG_LOG("Retaining old rule: '%s'", itcur->regex_text.c_str());
+      rules->push_back(*itcur);
+      ++itcur;
+    } else {
+      if (itload->expiry != itcur->expiry) {
+        DEBUG_LOG("Updating rule: '%s'", itload->regex_text.c_str());
+        rules->push_back(std::move(*itload));
+        changed = true;
+      } else {
+        DEBUG_LOG("Using old rule: '%s'", itcur->regex_text.c_str());
+        rules->push_back(*itcur);
+      }
+      ++itcur;
+      ++itload;
+    }
+  }
+
+  // any leftover loaded rules get tacked on
+  while (loaded.end() != itload) {
+    DEBUG_LOG("Adding new rule: '%s'", itload->regex_text.c_str());
+    rules->push_back(std::move(*itload));
+    ++itload;
+    changed = true;
+  }
+
+  DEBUG_LOG("Rules have been changed: '%s'", changed ? "true" : "false");
+
+  return changed;
+}
+
+// remove expired rules
+bool
+pruneConfig(time_t const timenow, std::vector<Invalidate> *const rules, time_t *const min_expiry)
+{
+  TSAssert(nullptr != rules);
+  TSAssert(nullptr != min_expiry);
+
+  bool pruned = false;
+  if (timenow < *min_expiry || nullptr == rules) {
+    return pruned;
+  }
+
+  // recalculate min_expiry
+  *min_expiry = std::numeric_limits<time_t>::max();
+
+  std::vector<Invalidate>::iterator itvec(rules->begin());
+
+  while (rules->end() != itvec) {
+    Invalidate const &inv = *itvec;
+
+    if (inv.expiry < timenow) {
+      DEBUG_LOG("Removing rule: %s", inv.regex_text.c_str());
+      std::swap(*itvec, rules->back());
```

Comment:
I think this needs to be:
```
if (itvec != rules->rbegin().base()) {
  *itvec = std::move(rules->back());
} else {
  itvec = rules->end();
}
```

Reply:
thanks for pointing this out. Actually this breaks some other code which needs the items to be sorted.

Reply:
Why can't this just be:
```
*itvec = std::move(rules->back());
```
?
https://godbolt.org/z/3rbfeefd4

Reply:
yuk missed this one as well.  have to erase to maintain sort order.

Reply:
Maybe instead of erasing you can just clear regex_text, then do:
```
auto iter1 = rules->begin();
auto iter2 = iter1;
while (iter1 != rules->end()) {
  if (!iter2->regex_text.empty()) {
    *(iter1++) = std::move(*iter2);
  }
  ++iter2;
}
rules->resize(iter1 - rules->begin());
```

Reply:
erase should already do that.  this is just occasionally loading it's not in the direct path of any transaction.